apps/dot_ext/tests/demographic_scopes_test_cases.py

```diff
@@ -0,0 +1,409 @@
+        "result_archived_token_count": 8,
+        "result_archived_data_access_grant_count": 5,
+    },
+    "test 13: app_requires = False bene_share = True request just PATIENT and A": {
+        # Request:
+        "request_app_requires_demographic": False,
+        "request_bene_share_demographic_scopes": True,
+        "request_scopes": SCOPES_JUST_PATIENT_AND_A,
+        # Result:
+        "result_has_error": False,
+        "result_token_scopes_granted": SCOPES_JUST_A,
+        "result_access_token_count": 1,
+        "result_refresh_token_count": 1,
+        "result_archived_token_count": 12,
+        "result_archived_data_access_grant_count": 6,
+    },
+    "test 14: app_requires = True bene_share = False request just PATIENT and A": {
+        # Request:
+        "request_app_requires_demographic": True,
+        "request_bene_share_demographic_scopes": False,
+        "request_scopes": SCOPES_JUST_PATIENT_AND_A,
+        # Result:
+        "result_has_error": False,
+        "result_token_scopes_granted": SCOPES_JUST_A,
+        "result_access_token_count": 1,
+        "result_refresh_token_count": 1,
+        "result_archived_token_count": 13,
+        "result_archived_data_access_grant_count": 7,
+    },
+    "test 15: app_requires = None bene_share = False request just PATIENT and A": {
+        # Request:
+        "request_app_requires_demographic": None,
+        "request_bene_share_demographic_scopes": False,
+        "request_scopes": SCOPES_JUST_PATIENT_AND_A,
+        # Result:
+        "result_has_error": False,
+        "result_token_scopes_granted": SCOPES_JUST_A,
+        "result_access_token_count": 1,
+        "result_refresh_token_count": 1,
+        "result_archived_token_count": 14,
+        "result_archived_data_access_grant_count": 8,
+    },
+    # Testing initial auth request has request_scopes = None
+    #   The POST to oauth2_provider:authorize must have scope in the payload.
+    "test 16: app_requires = True bene_share = True request scopes = None": {
+        # Request:
+        "request_app_requires_demographic": True,
+        "request_bene_share_demographic_scopes": True,
+        "request_scopes": None,
+        # Result:  NOTE: This case has empty scopes on POST!
+        "result_has_error": True,
+        "result_raises_exception": AssertionError,
+    },
+    # Testing case that produce ValueError exception
+    "test 17: app_requires = False bene_share = True request just PATIENT error=True": {
+        # Request:
+        "request_app_requires_demographic": False,
+        "request_bene_share_demographic_scopes": True,
+        "request_scopes": SCOPES_JUST_PATIENT,
+        # Result:  NOTE: Since Patient is not allowed,
+        #                this case results in having empty scopes on post auth!
+        "result_has_error": True,
+        "result_raises_exception": ValueError,
+        "result_exception_mesg": "Scopes must be set on post auth.",
+    },
```

Comment:
Not sure how much value it adds, but you could do the inverse test for `app_requires` and `bene_share` here to make sure it throws the same error.

Reply:
> Not sure how much value it adds, but you could do the inverse test for `app_requires` and `bene_share` here to make sure it throws the same error.

@njdister I like this. I just added it as test 18.